record/record.go

```diff
@@ -35,106 +38,30 @@ type MetricRecord interface {
 	SetCounterStartTime(ctx context.Context, start time.Time) error
 }
 
-// Name of the Datastore kind where metric records are stored.
-const kindName = "MetricRecords"
-
-// DatastoreMetricRecord defines a Datastore entity that is used to store status information about an imported metric.
-type DatastoreMetricRecord struct {
-	Name        string
-	Query       string
-	LastUpdate  time.Time // last time we wrote any points to SD.
-	LastAttempt time.Time // last time we attempted an update.
-	LastStatus  string
-
-	// CounterStartTime is used to keep start timestamp for cumulative metrics.
-	CounterStartTime time.Time
-}
-
-// NewDatastoreMetricRecord returns a Datastore-based metric record for a given metric name.
-func NewDatastoreMetricRecord(ctx context.Context, name, query string) (*DatastoreMetricRecord, error) {
-	r := &DatastoreMetricRecord{Name: name}
-	if err := r.load(ctx); err != nil {
-		return nil, err
-	}
-	r.Query = query
-	return r, nil
-}
-
-// Write metric data back to Datastore.
-func (m *DatastoreMetricRecord) write(ctx context.Context) error {
-	_, err := datastore.Put(ctx, m.key(ctx), m)
-	return err
-}
-
-// Load metric record state from Datastore.
-func (m *DatastoreMetricRecord) load(ctx context.Context) error {
-	err := datastore.Get(ctx, m.key(ctx), m)
-	if err != nil && err != datastore.ErrNoSuchEntity {
-		return err
-	}
-	return nil
-}
-
-// key returns the Datastore key for a given metric record.
-func (m *DatastoreMetricRecord) key(ctx context.Context) *datastore.Key {
-	return datastore.NewKey(ctx, kindName, m.Name, 0, nil)
-}
-
-// GetLastUpdate returns LastUpdate timestamp.
-func (m *DatastoreMetricRecord) GetLastUpdate() time.Time {
-	return m.LastUpdate
-}
-
-// GetCounterStartTime returns CounterStartTime.
-func (m *DatastoreMetricRecord) GetCounterStartTime() time.Time {
-	return m.CounterStartTime
-}
-
-// SetCounterStartTime sets CounterStartTime and persists metric data.
-func (m *DatastoreMetricRecord) SetCounterStartTime(ctx context.Context, start time.Time) error {
-	m.CounterStartTime = start
-	return m.write(ctx)
-}
-
-// UpdateError updates metric status in Datastore with a given error message.
-func (m *DatastoreMetricRecord) UpdateError(ctx context.Context, e error) error {
-	log.Errorf(ctx, "%s: %s", m.Name, e)
-	m.LastStatus = fmt.Sprintf("ERROR: %s", e)
-	m.LastAttempt = time.Now()
-	return m.write(ctx)
-}
-
-// UpdateSuccess updates metric status in Datastore with a given message.
-func (m *DatastoreMetricRecord) UpdateSuccess(ctx context.Context, points int, msg string) error {
-	log.Infof(ctx, "%s: %s", m.Name, msg)
-	m.LastStatus = fmt.Sprintf("OK: %s", msg)
-	m.LastAttempt = time.Now()
-	if points > 0 {
-		m.LastUpdate = time.Now()
+// NewStoredMetricRecord provides an abstraction for pluggable storage engines implementing MetricRecord interface
+func NewStoredMetricRecord(ctx context.Context, name, query string, storage string) (MetricRecord, error){
```

Comment:
Do you anticipate the need to use several different storage engines as part of the same ts-bridge deployment? If not, these wrappers seem unnecessary and a cleaner way might be to define a storage provider interface (it seems like it only needs to have two functions: `NewRecord` and `CleanupRecords`), make two implementations of it (one for Datastore, the other for BoltDB), and instantiate one of the two providers when the app is initialized.

Reply:
Gotcha. Abstracted the storage into a `StorageManager`, PTAL.